### PR TITLE
Allow passing arguments as options after the -- option

### PR DESCRIPTION
**Pull request by BerserkerTroll**
`root macro.C -- arg1 arg2 arg3 ...` will behave as `root macro.C(arg1,arg2,arg3,...)`.

Options won't be attached to:
- expressions — `root -e expression`;
- macros, passed with options (with `(` in them, to be precise) — `root macro.C(arg1, arg2)`;
- `.root` files;

If there are several macros without options, the arguments will be passed to the last one (with warning).
If there are no macros, the options after the `--` will be ignored (with warning).

No options description was updated yet.

Initially my idea was to allow to turn macro function parameters into named options, but @Axel-Naumann asked to leave them positional. But I'm still planning to introduce named arguments/options. Positional arguments as options is not a great improvement over the current way to pass arguments to macros.

### CODE REVIEW

File: core/base/src/TApplication.cxx [hunk 1]
```diff
@@ -468,7 +468,46 @@ void TApplication::GetOptions(Int_t *argc, char **argv)
...
+
+               if (file->TestBit(kExpression))
+                  continue;
+               if (file->String().EndsWith(".root"))
+                  continue;
+               if (strchr(file->String().Data(), '('))
+                  continue;
+
+               if (macro)
+                  Warning("GetOptions", "-- is used with several macroses. "
```
Comment (Axel-Naumann) [inline core/base/src/TApplication.cxx]: Spelling: two times a macro is two "macros" :-)

File: core/base/src/TApplication.cxx [hunk 2]
```diff
@@ -468,7 +468,46 @@ void TApplication::GetOptions(Int_t *argc, char **argv)
...
 
+         if (fFiles) {
+            for (int f = 0; f < fFiles->GetEntriesFast(); f++) {
+               TObjString* file = dynamic_cast<TObjString*>(fFiles->At(f));
+
+               if (file->TestBit(kExpression))
+                  continue;
+               if (file->String().EndsWith(".root"))
+                  continue;
+               if (strchr(file->String().Data(), '('))
```
Comment (Axel-Naumann) [inline core/base/src/TApplication.cxx]: Don't we have a TString function which can replace the `strchr`? Like `Index()` or something?

File: core/base/src/TApplication.cxx [hunk 3]
```diff
@@ -468,7 +468,46 @@ void TApplication::GetOptions(Int_t *argc, char **argv)
          } else {
             Warning("GetOptions", "-e must be followed by an expression.");
          }
+      } else if (!strcmp(argv[i], "--")) {
+         TObjString* macro = NULL;
```
Comment (Axel-Naumann) [inline core/base/src/TApplication.cxx]: Can you use `nullptr` instead of `NULL`? (I know, not obvious when looking at surrounding code...)

File: core/base/src/TApplication.cxx [hunk 4]
```diff
@@ -468,7 +468,46 @@ void TApplication::GetOptions(Int_t *argc, char **argv)
...
+               if (macro)
+                  Warning("GetOptions", "-- is used with several macroses. "
+                                        "The arguments will be passed to the last one.");
+
+               macro = file;
+            }
+         }
+
+         if (macro) {
+            argv[i] = null;
```
Comment (Axel-Naumann) [inline core/base/src/TApplication.cxx]: Should this `null` be `nullptr`?

**Comment by Axel-Naumann** [comment]
@phsft-bot build!

File: core/base/src/TApplication.cxx (hunk 1 above)
Comment (BerserkerTroll) [inline core/base/src/TApplication.cxx]: May I amend-commit?

File: core/base/src/TApplication.cxx [hunk 5]
```diff
@@ -468,7 +468,46 @@ void TApplication::GetOptions(Int_t *argc, char **argv)
...
+            ++i;
+            TString& str = macro->String();
+
+            str += '(';
+            for (; i < *argc; i++) {
+               str += argv[i];
+               str += ',';
+               argv[i] = null;
+            }
+            str.EndsWith(",") ? str[str.Length()-1] = ')' : str += ')';
```
Comment (Axel-Naumann) [inline core/base/src/TApplication.cxx]: Except for adding spaces around the "-" in `Length() - 1`, I don't think clang-tidy's suggestions are of much value. So please ignore its "failure".

File: core/base/src/TApplication.cxx (hunk 4 above)
Comment (BerserkerTroll) [inline core/base/src/TApplication.cxx]: No, it shoud not, because later options participate in `strcmp` ([1]), which doesn't like null pointers. `null` is defined as `""`.

[1] https://github.com/BerserkerTroll/root/blob/385d7c5f7196c9b47f8dbc6d126bc6cd079d7dec/core/base/src/TApplication.cxx#L592

File: core/base/src/TApplication.cxx (hunk 1 above)
Comment (Axel-Naumann) [inline core/base/src/TApplication.cxx]: Anything that's not a merge commit is perfectly fine: amend, rebase, add commits - whatever works for you!

File: core/base/src/TApplication.cxx [hunk 6]
```diff
@@ -468,7 +468,48 @@ void TApplication::GetOptions(Int_t *argc, char **argv)
          } else {
             Warning("GetOptions", "-e must be followed by an expression.");
          }
+      } else if (!strcmp(argv[i], "--")) {
+         TObjString* macro = nullptr;
+
+         if (fFiles) {
+            for (auto f = fFiles->GetEntriesFast(); f --> 0; ) {
```
Comment (Axel-Naumann) [inline core/base/src/TApplication.cxx]: Ha! Funny syntax :-) Nonetheless - could you reword this to the more common (for ROOT)
```
for (auto f: *fFiles)
```

File: core/base/src/TApplication.cxx (hunk 4 above)
Comment (Axel-Naumann) [inline core/base/src/TApplication.cxx]: Thanks, indeed, you're right!

**Comment by Axel-Naumann** [comment]
Could you (in a separate PR) hand in a test for this wonderful feature to the roottest repo?

Could you also add a few lines about this to README/ReleaseNotes/v612/index.md?

**Comment by pcanal** [comment]
We still also need some documentation for it (similar to the PR description at least).

**Comment by Axel-Naumann** [comment]
@phsft-bot build!

**Comment by Axel-Naumann** [comment]
We seem to have two successful builds versus three infrastructure failures - so I suppose that's "green". Merging - thank you for your contribution!
